tools/codegen/api/python.py

```diff
@@ -0,0 +1,955 @@
+    # as understood by mypy; our format was independently developed
+    # and has some quirks to make it more suitable specifically
+    # for error parsing.
+    #
+    # For a translation to mypy-valid type signatures, see
+    # tools/gen_pyi.py.  If you change any logic here, please
+    # check that file too.
+    def signature_str(self, *, skip_outputs: bool = False) -> str:
+        schema_formals: List[str] = \
+            list(map(lambda a: a.argument_str(method=self.method),
+                     self.arguments(skip_outputs=skip_outputs)))
+        positional_argc = len(self.input_args)
+        if len(schema_formals) > positional_argc:
+            schema_formals.insert(positional_argc, '*')
+
+        return f'{self.name}({", ".join(schema_formals)})'
+
+# The deprecated python signature involves some special logic, so create a
+# dedicated data model to store these extra properties.
+
+
+@dataclass(frozen=True)
+class PythonSignatureDeprecated(PythonSignature):
+    # We need keep the order of arguments in deprecated signature.
+    # Particularly, method signature might have 'self' not at the beginning, e.g.:
+    #   addmm(Scalar beta, Tensor self, Tensor mat1, Tensor mat2)
+    # When generating lambda function signature we need follow the exact order (even for method=True):
+    #   [](Scalar beta, const Tensor & self, const Tensor & mat1, const Tensor & mat2) -> Tensor
+    deprecated_args_names: Tuple[str]
+
+    # The deprecated signature might miss some arguments that the corresponding
+    # C++ signature expects. We need store the constant default values to pass in.
+    # For example:
+    #   [deprecate signature]: addmm(Scalar beta, Tensor self, Tensor mat1, Tensor mat2)
+    #   [func schema]: aten::addmm(Tensor self, Tensor mat1, Tensor mat2, *, Scalar beta=1, Scalar alpha=1) -> Tensor
+    #   [func call]: self.addmm(mat1, mat2, beta, 1)
+    # We store ['self', 'mat1', 'mat2', 'beta', '1'] in this case.
+    deprecated_args_exprs: Tuple[str]
+
+    @property
+    def deprecated(self) -> bool:
+        return True
+
+    def signature_str(self, *, skip_outputs: bool = False) -> str:
+        return PythonSignature.signature_str(self, skip_outputs=skip_outputs) + '|deprecated'
+
+# C++ function dispatch is wrapped in a lambda function. The lambda function
+# has almost the same signature as the C++ function, only with some small
+# variants - see details below.
+# This data model is used to represent arguments of the lambda function
+# signature.
+
+
+@dataclass(frozen=True)
+class DispatchLambdaArgument:
+    name: str
+    type_str: str
+    cpp_type_str: str
+    is_out_arg: bool
+
+# To pass PyObjects arguments to C++ function (via the lambda wrapper),
+# we need first convert PyObjects into simple C++ objects. This work
+# is done by PythonArgParser.
+# This data model is used to represent the output of PythonArgParser.
+# It has 1-1 mapping with PythonArgument in PythonSignature.
```

Comment:
Hot take: this should just be CppArgument. The reasoning here is that the redispatch logic in Python argument parser goes through the cpp API (e.g., at::add(...)). So the conversion going on here is PythonArgument ~> CppArgument. (This also suggests that maybe we should go straight to DispatcherArgument rather than via the intermediary CppArgument, but that wouldn't be byte-for-byte compatible so OK not to do here.)

Reply:
(this = DispatchLambdaArgument)

Reply:
> Hot take: this should just be CppArgument. The reasoning here is that the redispatch logic in Python argument parser goes through the cpp API (e.g., at::add(...)). So the conversion going on here is PythonArgument ~> CppArgument.

Right - I started with the same CppArgument but changed my mind because this lambda signature has some special conversions compared to CppSignature (check out the comments above `dispatch_lambda_args`), main about passing-by-value v.s. passing-by-reference.
So I created this type because technically the lambda wrapper here can diverge from the DispatcherArgument at a distant place...